### PR TITLE
mon,osd: use GMT time for the object name of hitsets

### PR DESCRIPTION
**Pull request by tchaikov**
http://tracker.ceph.com/issues/9732 Fixes: #9732

### CODE REVIEW

**Comment by tchaikov** [comment]
was #4941

**Comment by liewegas** [comment]
would it make sense to allow 'ceph osd pool set $pool use_gmt_hitset' when hit_set_count == 0?

**Comment by tchaikov** [comment]
> would it make sense to allow 'ceph osd pool set $pool use_gmt_hitset' when hit_set_count == 0?

actually, thanks to the granularity of `pg_hit_set_info_t.using_gmt`, `hit_set_count == 0` is not necessary, i think. we can safely trim/read the archived hit set archives using the gmt or localtime by looking at `pg_hit_set_info_t.using_gmt`.

**Comment by liewegas** [comment]
s/gmt/utc/ ?

**Comment by tchaikov** [comment]
@liewegas gmt might be more helpful for the reader, as we are basically switching to `gmtime_r(3)` in this change.

```patch
+  if (using_gmt) {
+    start.gmtime(ss) << "_";
+    end.gmtime(ss);
+  } else {
+    start.localtime(ss) << "_";
+    end.localtime(ss);
+  }
```

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -1834,6 +1835,20 @@ bool OSDMonitor::preprocess_boot(MonOpRequestRef op)
...
   }
 
+  if (any_of(osdmap.get_pools().begin(),
+	     osdmap.get_pools().end(),
+	     [](const std::pair<int64_t,pg_pool_t>& pool)
+	     { return pool.second.use_gmt_hitset; })) {
+    assert(osdmap.get_num_up_osds() == 0 ||
+	   osdmap.get_up_osd_features() & CEPH_FEATURE_OSD_HITSET_GMT);
+    if (!(m->osd_features & CEPH_FEATURE_OSD_HITSET_GMT)) {
+      dout(0) << __func__ << " conf requires 'osd hitset gmt' but osd at "
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: "one or more pools uses GMT hitsets but ..." ?  it's not the conf we're checking here

**Comment by liewegas** [comment]
aside from that warning msg nit, lgtm!  Reviewed-by:.  add needs-qa?

File: src/mon/OSDMonitor.cc (hunk 1 above)
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: thanks, will do.

**Comment by tchaikov** [comment]
updated the warning message as suggested by sage and repushed.

**Comment by tchaikov** [comment]
needs to run upgrade suite from hammer when doing cache/tiering

**Comment by tchaikov** [comment]
the test added at https://github.com/ceph/ceph-qa-suite/tree/wip-tiering-hammer-x , will test it tomorrow.

**Comment by tchaikov** [comment]
should reuse the `CEPH_FEATURE_OSD_BITWISE_HOBJ_SORT` bit, but with another name.

as this pr will be merged along with https://github.com/ceph/ceph/commit/73b3ed87c16fa66c9e2f77b9dadcb2615a118201#diff-33b7fd478ccf759677a0e515638a9a81R67

**Comment by tchaikov** [comment]
rebased against master to resolve merge conflicts and repushed.

**Comment by tchaikov** [comment]
rebased & repushed

**Comment by liewegas** [comment]
Final thought.. if we think there is any possibility we'll backport this feature to hammer then we should *not* reuse the hobj sort feature bit... @athanatos ?

**Comment by tchaikov** [comment]
rebased against master to fix the build failure

**Comment by liewegas** [comment]
passed my wip-sage-testing run, sage-2015-08-21_04:15:01-rados-wip-sage-testing-distro-basic-multi

**Comment by tchaikov** [comment]
fix the compat version of `pg_hit_set_info_t::encode()`, it should be 1

**Pull request merged**